File: leetcode/062-unique-paths.clj
#!/usr/bin/env lein-exec


(defn ones [n]
  (take n (repeat 1)))

(defn add-top-and-left [xs y]
  (concat xs [(+ (last xs) y)]))

(defn next-row [prior _]
  (rest (reduce add-top-and-left [0] prior)))

(defn unique-paths [width height]
  (last (reduce next-row (ones (+ 1 width)) (range height))))


(println (unique-paths 3 4))  ;; 35

File: leetcode/062-unique-paths.hs
#!/usr/bin/env runghc

ones n = take n $ repeat 1
next prior = 0 : [p + n | (p, n) <- zip prior (next prior)]
nextRow prior _ = tail $ next prior
uniquePaths width height = last $ foldl nextRow [1..height] (ones width)

main = print $ uniquePaths 3 4 -- 35
